### PR TITLE
[Python AMQP] Link Error on Detach Fix

### PR DESCRIPTION
**Pull request by kashifkhan**
The clients currently output an error message when the sender / receiver clients close `An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.InternalError\n Error Description: Link already closed.')`

### Why this happens
client.py ( and async ) have a reference to a link and a CBSAuthenticator object (which has 2 links via management link, `request_link` & `response_link`

When a client is closing the following happen
* The client first closes its own link. Simple detach
* Then CBSAuth is closed, which detaches the management link [here](https://github.com/Azure/azure-sdk-for-python/blob/1abca3a9a479d15fb660e912a9a0256b2cf3590d/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/management_link.py#L248)
  * The close here sets management link to idle and errors out any pending operations on this link 
* Finally the session is ended
  * This step also tries to detach every link created on it. As its closing links that have already been closed above, errors are raised by the `_check_if_closed` that the link is already closed. 

Initially I thought about letting only the session detach links on end and working my way backwards ( detach links, end session and close connection). The closing of the management link and the state change step made that problematic as management would start erroring out. The management link state was not the proper one and changing it from the client looked wrong. 

The simplest way to go about this was checking if the links are already closed during session end.

### CODE REVIEW

**Review by l0lawrence — COMMENTED** [review]
Can we update SB as well in this PR just so they are aligned?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py [hunk 1]
```diff
@@ -414,7 +414,8 @@ async def end(self, error=None, wait=False):
             if self.state not in [SessionState.UNMAPPED, SessionState.DISCARDING]:
                 await self._outgoing_end(error=error)
             for _, link in self.links.items():
-                await link.detach()
+                if link._is_closed is False: # pylint: disable=protected-access
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py:417]: Looking at the link detach code, it looks like `close=False` by default, which means calling `link.detach()` without parameters wont result in `is_closed` being set to True.
So it seems that the original intention behing `is_closed` was _not_ to be used interchangably with `is_detached`, which is how it's used here.

Comment (annatisch) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py:417]: Whether this requires a more extensive audit of the use of `_is_closed` or not, I'm not sure.
It is an internal attribute to `link` - so should not be consumed by Session in this manner. Though I'm not sure if anything else is also consuming it like this.
If we do need to expose some kind of access to the Session object, we should add a public attribute/method for it.

Comment (kashifkhan) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py:417]: yeah that was the conundrum as `_is_closed` is used even at the client [level](https://github.com/Azure/azure-sdk-for-python/blob/6f4abdd4a870a9ef33981d9c3ca7d18ceffdbc75/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/client.py#L234). 

I was wondering instead of adding an attribute on `Link`, adding something like `detach_link(link_name)` method on Session. The links created by client and mgmt are done via `Session` methods on Session (`create_receiver_link`, `create_sender_link`, `create_request_response_link_pair`) , so detach should also be done at that level ?

Comment (kashifkhan) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py]: thankfully its usage is outside of `link` is limited to `session`, `client` and `receiver` and that too sparingly.  Given that service bus doesn't support link recovery, I'm guessing the use was to determine if the link was in a detaching state and not whether its closed. Ive added a `is_detaching` method and maybe `is_closed` or ( stopped ) is something to expose for the future.

**Comment by kashifkhan** [comment]
Update:

After a relook, the changes now include updating Link.detach() to not send a detach when the `LinkState` is already detached or a detach has been sent (server has not responded yet).

Valid states to send a detach -> Attach, Attach Sent and Link Error. Per the spec ( in comments ) if a link is in error state the endpoint must be destroyed and detached. In the case of python amqp, we raise an error at the `ReceiverLink` level. 

`check_if_closed` has been removed from the detach call. close is used to determine if a link can be reattached, which means the link is in a detached state.

**Review by swathipil — COMMENTED** [review]
left a few small comments, but o/w seems reasonable to me

File: sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/link.py [hunk 2]
```diff
@@ -241,11 +241,10 @@ def attach(self):
         self._set_state(LinkState.ATTACH_SENT)
 
     def detach(self, close=False, error=None):
-        if self.state in (LinkState.DETACHED, LinkState.ERROR):
+        if self.state in (LinkState.DETACHED, LinkState.DETACH_SENT, LinkState.ERROR):
             return
         try:
-            self._check_if_closed()
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/link.py:247]: If we had previously received an incoming Detach due to an error scenario, that error would be assigned to `self._error` on line 232, which would then be raised here (though this had been incorrectly convoluted with `self._is_closed`.
By removing this line, where will that error ultimately get propagated? And does this have flow-on effects? 
You mention in your comment above that any errors have already been raised by this point - where?

Comment (kashifkhan) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/link.py:247]: There are 2 places which would be immediate, one at the receiver `[_wait_for_response]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/receiver.py#L79) , and the other at the sender by calling `[_check_if_closed]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/sender.py#L164)

I think the above two flows are also slightly different due to the confusion between `is_closed` and detached. The errors would be raised here as they try to perform a send or receive.

Another way errors in `self._error` would be raised is via `link.get_state` which is called in `[client.py]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/client.py#L550) ( and is part of the bigger while loop of `do_work` and  `client_ready` in the same file )

### Flow impact

`self._error` at the link level only gets populated `_incoming_detach` on line 232. After this the link can be in two states 
Detached or Error (for this case point self._error is populated)

both of these states are in the `if state in return` block of detach. They would never raise there.

if a link is any `ATTACH*` or `DETACH_SENT` state ( current implementation), `self._error` would not have anything. 

Further, in Attach state `is_closed` would be False and would break out from `check_if_closed`.

Detach_sent would be only time `is_closed` is true, but thats what would cause the `Link Error on Detach` to trigger, which is incorrect, we dont have any information from the service yet.  

So currently `check_if_closed` being called here would have an effect for 1 state. 

with the new change, if state is DETACH_SENT, no need to send a detach and at some point a detach would come from the service with any errors.

**Review by annatisch — APPROVED** [review]

**Comment by Benniah** [comment]
Hi @kashifkhan 

I am running into this exact issue within my EventHubConsumerClient.
I don't have this issue when i switch to **uamqp_transport=True** 

**some useful info :** 

Package Name: azure-eventhub
Package Version: 5.12.0
Operating System:linux
Python Version: 3.9.1
